### PR TITLE
pat_env points to the correct environment

### LINKED ISSUE #9799: Matching against GADT tuple when multiple patterns result in the same expression produces an error
I'm not sure if it's a bug or by design but match in `fn` produces an error:

```ocaml
type 'a t =
  | A: [ `a  | `z ] t
  | B: [ `b  | `z ] t

let fn: type a . a t -> a -> int = 
  fun x y ->
    match (x, y) with
    | (A, `a)
    | (B, `b) -> 0
    | (A, `z)
    | (B, `z) -> 1
```

```
Error: This pattern matches values of type [ `a | `z ] t
       but a pattern was expected which matches values of type a t
       Type [ `a | `z ] is not compatible with type a
```

P.S. I'm building against `ocaml: ~4.6.1000`

### PR DESCRIPTION
**Pull request by trefis**
That is: the one actually used to type the pattern, not the one used to type one of its "ancestors".

This fixes #9799 and does the opposite of @gasche's request in #9081: it adds a test that will fail if the pattern environment is changed after type-checking.

### CODE REVIEW

**Comment by gasche** [comment]
Indeed, we didn't insist to have a discussion about what was happening at the time, and we probably should have.

**Comment by gasche** [comment]
cc @garrigue: I discussed this again with @trefis today and it seems reasonable. Would you by chance remember what pat_env are made constant in this way, and whether this is still useful today?

**Comment by xavierleroy** [comment]
Any progress on this one?

**Comment by gasche** [comment]
In absence of further information from @garrigue, I think the better choice is to merge this PR. It simplifies the codebase and fixes a bug, and we don't know when and why the code removed was necessary in the first place. Sigh... I wish we could rely on better information to evolve the codebase.

**Comment by gasche** [comment]
@trefis you left trailing whitespace in the testsuite! This would be good to go when `check-typo` and the CI is happy.

**Comment by trefis** [comment]
@gasche I removed the trailing whitespace, and added you as reviewer in the Changes file.

**Comment by gasche** [comment]
I sort of regret this "merge-me" label. I don't get to merge pull requests anymore, everyone else is quicker.

**Comment by trefis** [comment]
Despite that `merge-me` label, I don't dare clicking on the merge button. There wasn't an official approval, has the protocol changed? :)

**Review by gasche — APPROVED** [review]

**Pull request merged**
